src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -499,267 +453,299 @@ test_budgets(void)
-	PRINTC(" Test - Intr = H, Worker = M, Preempted = L");
-	test_wakeup_case(&wat[cos_cpuid()], TEST_PRIO_LOW, TEST_PRIO_HIGH, TEST_PRIO_MED, -1, 0, 1);
+    //switch
+	EXPECT_LL_NEQ(0, pending, "Test Async Endpoints");
 
-	/*
-	 * test case 2: worker_thd = H, intr_thd = M, preempted_thd = L
-	 * expected result:
-	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
-	 * - cos_rcv from intr_thd should activate worker_thd.
-	 */
-	PRINTC(" Test - Worker = H, Intr = M, Preempted = L");
-	test_wakeup_case(&wat[cos_cpuid()], TEST_PRIO_LOW, TEST_PRIO_MED, TEST_PRIO_HIGH, -1, 0, 1);
+	pending = cos_rcv(rc, RCV_NON_BLOCKING, NULL);
+	EXPECT_LL_NEQ(pending, -EAGAIN, "Test Async Endpoints");
 
-	/*
-	 * test case 3: intr_thd = H, preempted_thd = M, worker_thd = L
-	 * expected result:
-	 * - cos_thd_wakeup should not invalidate preempted_thd.
-	 * - cos_rcv from intr_thd should activate preempted_thd.
-	 */
-	PRINTC(" Test - Intr = H, Preempted = M, Worker = L");
-	test_wakeup_case(&wat[cos_cpuid()], TEST_PRIO_MED, TEST_PRIO_HIGH, TEST_PRIO_LOW, 1, 0, -1);
+	pending = cos_rcv(rc, 0, NULL);
 
-	/*
-	 * test case 4: intr_thd = H, worker_thd = M, preempted_thd = L
-	 * 		(+ budget with cos_thd_wakeup)
-	 * expected result:
-	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
-	 * - cos_rcv from intr_thd should activate worker_thd.
-	 * - worker thread should be preempted after (budget==timeout) and that should activate it's scheduler.
-	 */
-	PRINTC(" Test Wakeup with Budget -");
-	wakeup_budget_test[cos_cpuid()] = 1;
-	seq_expected_order_set(&wat[cos_cpuid()].s, 2); /* scheduler should be activated for worker timeout */
-	final_seq[cos_cpuid()] = 3;
-	test_wakeup_case(&wat[cos_cpuid()], TEST_PRIO_LOW, TEST_PRIO_HIGH, TEST_PRIO_MED, -1, 0, 1);
+    //switch
+	EXPECT_LL_NEQ(0, 1, "Test Async Endpoints");
 
-	wakeup_test_start[cos_cpuid()] = 0;
-	PRINTC("Done.\n");
+	cos_thd_switch(tc);
+	EXPECT_LL_NEQ(0, 1, "Test Async Endpoints");
+	while (1) cos_thd_switch(tc);
 }
 
 static void
-receiver_thd(void *d)
+async_thd_parent(void *thdcap)
 {
-	struct exec_cluster *e = &(((struct activation_test_data *)d)->w);
+	thdcap_t    tc = (thdcap_t)thdcap;
+	arcvcap_t   rc = rcp_global[cos_cpuid()];
+	asndcap_t   sc = scp_global[cos_cpuid()];
+#if defined(PERF)
+	int	    pending;
+#endif
+	int         ret;
+	thdid_t     tid;
+	int         blocked, rcvd;
+	cycles_t    cycles, now;
+	tcap_time_t thd_timeout;
 
-	while (1) {
-		cos_rcv(e->rc, 0, NULL);
-		seq_order_check(e);
-	}
+	ret = cos_asnd(sc, 0);
+	ret = cos_asnd(sc, 0);
+	ret = cos_asnd(sc, 0);
+	ret = cos_asnd(sc, 1);
+
+    //switch
+	/* child blocked at this point, parent is using child's tcap, this call yields to the child */
+	ret = cos_asnd(sc, 0);
+
+    //switch
+	ret = cos_asnd(sc, 0);
+	EXPECT_LL_NEQ(0, ret, "Test Async Endpoints");
+
+    //switch
+	ret = cos_asnd(sc, 1);
+	EXPECT_LL_NEQ(0, ret, "Test Async Endpoints");
+
+    //switch
+#if defined(PERF)
```

Comment:
Looks like this is not required if you cleaned up to get rid of PRINTC below.

Reply:
FIXED